docs/IR.md

```diff
@@ -237,7 +237,7 @@ A name belonging to the Value namespace may appear in multiple places, namely as
 
 A value name used in a graph must have a unique definition, with the exception that the same name MAY appear in both the graph input list and graph initializer list. (Further exceptions apply in the presence of nested subgraphs, as described later.)
 
-When a name appears in both the initializer list and the graph input list, a runtime MAY allow a caller to specify a value for this (input) name overriding the value specified in the initializer and a runtime MAY allow users to omit specifying a value for this (input) name, choosing the value specified in the initializer. Names of constants that are not meant to be overridden by the caller should appear only in the initializer list and not in the graph input list. In nested subgraphs used as attribute values, users MUST NOT use the same name as both a subgraph initializer and subgraph input unless the corresponding op's specification explicitly allows it.
+When a name appears in both the initializer list and the graph input list, a runtime MAY allow a caller to specify a value for this (input) name overriding the value specified in the initializer and a runtime MAY allow users to omit specifying a value for this (input) name, choosing the value specified in the initializer. Names of constants that are not meant to be overridden by the caller should appear only in the initializer list and not in the graph input list. In models with IR version >= 4, in nested subgraphs used as attribute values, users MUST NOT use the same name as both a subgraph initializer and subgraph input unless the corresponding op's specification explicitly allows it. In models with IR version <= 3, users MAY use the same name as both a subgraph initializer and subgraph input. Note that in models with IR version <= 3, the actual inputs expected by the control-flow operator and supplied by the node containing the invocation of the control-flow operator must appear as the initial inputs of the subgraph, with extraneous initializers following subsequently, so that the actual inputs and formal inputs can be matched positionally.
```

Comment:
I like the clarification, but for IR version <= 3 this is not 100% clear as to whether inputs with initializers (optional inputs) can be used as initial, actual inputs (provided they line up correctly).  If you are disallowing it, you should maybe change
`the actual inputs expected by the control-flow operator and supplied by the node containing the invocation of the control-flow operator must appear as the initial inputs of the subgraph, with extraneous initializers following subsequently, so that the actual inputs and formal inputs can be matched positionally.`
to
`the actual inputs expected by the control-flow operator and supplied by the node containing the invocation of the control-flow operator cannot have initializers and must appear as the initial inputs of the subgraph, so that the actual inputs and formal inputs can be matched positionally. Extraneous initializers may follow subsequently.`

Reply:
How about changing the two sentences to
```
In models with IR version <= 3, users MAY use the same name as both a subgraph initializer and subgraph input,
but this is restricted to support constants via initializers that are not intended to correspond to any
actual inputs expected by the control-flow operator and supplied by the node containing the invocation
of the control-flow operator. In particular, the control-flow operator semantics determines the set of inputs
supplied to the execution of the subgraph, and these input names MUST NOT appear as subgraph initializers.
Subgraph initializer names must appear in the graph input list _after_ the actual inputs. This allows the actual
inputs and formal inputs to be matched positionally.
```

Reply:
Yes that's good.  nit: I'd revise `not intended to correspond to any
actual inputs expected by the control-flow operator and supplied by the node containing the invocation of the control-flow operator.` 
to 
`not intended to correspond to any actual inputs passed from the node into the subgraph.`

Reply:
I took both Rama's new sentences and Tom's revision since it looks clearer to me